### DISCUSSION TITLE
Questions on Common Lisp in Medley

### DISCUSSION THREAD

**Discussion by masinter** [Interlisp/medley#1473] [General]
u/Lispm on Reddit asked these questions, which I think deserve a considered answer. There are other 'common lisp' discussions and issues, but this one is focused on answering these specific questions.
=================

But specifically about CL... Medley/Interlisp-D was developed originally developed in Interlisp, which is quite different from Common Lisp in many ways. A version of Common Lisp was then integrated, which then runs side-by-side with Interlisp. I don't know if it was actually used by Medley users or if the users mostly kept using Interlisp.

What is the overall state of the Common Lisp compatibility of Medley?

How much of the standard CL language runs? CLOS? Conditions? LOOP? Pretty Print?

Does it run things like ASDF, Quicklisp, Slime, base libs?

How does it integrate its image-based source control with the typical file/system based source model (used in most other Common Lisps)? Can a Common Lisp developer use Medley to develop CL code as productive as in other systems (say, SBCL/GNU-Emacs/SLIME?) or, possibly, be more productive? How does one share code and libraries?

**Comment by masinter** [comment]
# Short answer

We don't know. Would you like to help us find out?

# longer answer

> But specifically about CL... Medley/Interlisp-D was developed originally developed in Interlisp, which is quite different from Common Lisp in many ways. 

ok

> A version of Common Lisp was then integrated,

bit by bit

> then runs side-by-side with Interlisp

they're much more intertwined than 'side-by-side' implies.

> I don't know if it was actually used by Medley users or if the users mostly kept using Interlisp.

T have heard at least two people claim. and see some evidence that Medley might have been used by them to develop Common Lisp code. 

> How much of the standard CL language runs?

Hard to measure -- what is the scale? Enough to use?  It claims "CLtL1+". I don't know.
While I was active in X3J13, I stopped working on Common Lisp in 1986. Lots of changes were made since.

> CLOS? 

apparently. in PCL. See the clos folder in the Medley repo.

> Conditions?

apparently, but not certain. Implemented using "spaghetti stacks"

>  LOOP? 

Interlisp has its own edition of this. Anyway, LOOP isn't there and isn't in CLTL1.

> Pretty Print?

Package-aware pretty printing is an essential feature of the file manager and SEdit.
 
> Does it run things like ASDF, Quicklisp, Slime, base libs?

The answer is "no" and any one of those will be difficult because Medley has its own significantly different perspective on loading and libraries. I don't know what 'base libs' are.

> How does it integrate its image-based source control with the typical file/system based source model (used in most other Common Lisps)?

It's probably misleading to call Medley "image based". Yes there are images and facilities for making them, and some applications were packaged as images, but for the most part Medley deals in "files" but from a different perspective. (I've heard this called 'repl-driving development'. When convenient, you write out the repl-entered definers into a "file" and compile that. If you want to upgrade your base, you can then load the compiled and/or interpreted code in your new base image.

There is some code I haven't tried to convert both ways the "managed file" view and the "text editor source" view.

> ? Can a Common Lisp developer use Medley to develop CL code as productive as in other systems (say, SBCL/GNU-Emacs/SLIME?) or, possibly, be more productive?

That's the big question -- and there won't be a simple answer. Back in the day, people without a lot of programming experience could learn and use Interlisp quickly. The world has changed. 

> ? How does one share code and libraries?

lots of examples in the Medley repo under /library and /lispusers of distributed source code, library and documentation.

**Comment by pamoroso** [comment]
> I don't know if it [Common Lisp] was actually used by Medley users or if the users mostly kept using Interlisp.

The SpinPro expert system was cross developed in Common Lisp on Interlisp-D and deployed to MS-DOS via Golden Common Lisp by Gold Hill. @MattHeffron worked on SpinPro. For more details see [Volume 1, Number 3 of MASTERSCOPE](http://www.bitsavers.org/pdf/xerox/interlisp-d/newsletters/Masterscope_1-03_Aug85.pdf) (August 1985) on page 2 of the PDF file.

**Reply by MattHeffron** [reply]
A little expansion on what @pamoroso wrote: SpinPro (and a subsequent expert system, PepPro) were developed with **targeting Common Lisp for deployment**, but the in-house (at Beckman Instruments) Expert System _development environment_ was written in Interlisp, taking advantage of the Interlisp-D environment.

**Comment by Anzus** [comment]
At OSU we were developing CL code in Medley. The Generic Tasks Toolset (https://ntrs.nasa.gov/citations/19890017176), which was the framework for other languages for GT, was all in CL. The other developer on the project was using a TI Explorer so we could ensure compatibility with other CL implementations,  the grad students primarily used d-machines. I think at least a couple of dissertations came out of it. We had to keep compatibility with other CL’s because grad students would go on to other universities  and research labs and wanted to use it on whatever systems they ended up with (at that time they were mostly transitioning to Franz  and Lucid on Sun, it was getting harder to get grant money for dedicated Lispm’s). CLtL1 was pretty portable between implementations.

**Comment by tfeb** [comment]
On 22 Dec 2023, at 06:59, Larry Masinter ***@***.***> wrote:
> 
> I have heard at least two people claim. and see some evidence that Medley might have been used by them to develop Common Lisp code.

I don't know if one of those people was me, but you can add me to that set.  I used Medley on dmachines in about 1989-1992, and wrote almost entirely CL code.  I still maintain descendants of some of that code today.

This was all pre ANSI CL of course: in particular things like loop and CLOS (in the form of PCL) were things you fetched and built.  MIT loop was I think what everyone used and I can still remember some of the PCL release names: rainy day, Victoria day (was that the last one?).

I haven't explored the Medley-revival environments although I should, but Medley was a very competent, if somewhat strange, CL environment when I used it.

--tim

**Comment by pamoroso** [comment]
Is  `loop` really supposed to be missing from Medley's Common Lisp implementation?

Some of the code snippets in the [Medley 1.0 release notes](http://www.bitsavers.org/pdf/xerox/interlisp-d/198809_Medley_1.0/400006_Lisp_Release_Notes_Medley_Release_1.0_Sep88.pdf) contain `loop` forms, for example on page 237 of the PDF:

![loop1](https://github.com/Interlisp/medley/assets/10342479/1f4343bb-d9ef-4c83-9373-a84edb093131)

Although there's apparently a relevant definition of `loop`, the expressions I tried return errors or nothing happens when executed:

![loop2](https://github.com/Interlisp/medley/assets/10342479/7d3c80db-0620-4932-9135-117050e90e55)

The only expression I've been able to execute correctly so far is `(loop (print "Infinite loop."))`:

![loop3](https://github.com/Interlisp/medley/assets/10342479/ca384a4d-b75f-4877-975e-9e9b4d628aa0)

**Comment by Anzus** [comment]
Medley is CLtL1, which predates LOOP.On Jan 29, 2024 5:09 AM, Paolo Amoroso ***@***.***> wrote:
Is  loop really supposed to be missing from Medley's Common Lisp implementation?
Some of the code snippets in the Medley 1.0 release notes contain loop forms, for example on page 237 of the PDF:
loop1.png (view on web)
Although there's apparently a relevant definition of loop, the expressions I tried return errors or nothing happens when executed:
loop2.png (view on web)
The only expression I've been able to execute correctly so far is (loop (print "Infinite loop.")):
loop3.png (view on web)

**Comment by nbriggs** [comment]
The definition for LOOP in the current (CLTL1) code is in sources/CMLSPECIALFORMS, and it just expands to a basic PROG with an infinite loop that you can RETURN from (or error out of) - there's none of the CLTL2 fancy stuff there.

**Reply by pamoroso** [reply]
So it's barely a stub. Interestingly, just enough to make the sample code in the Medley 1.0 release notes work.

**Comment by masinter** [comment]
I believe the spec for Common Lisp loop was finalized late in the process. 
I've looked (not very hard or successfully) for a definition of ANSI CL LOOP that Medley could use. There might be one in an open source Common Lisp.  SBCL? SICL? Translate CL:LOOP into Interlisp iterative statements? 

One thing that I've been thinking about is the package used for lexically-scoped variables might be more robust, more readble, if they didn't show up in the "XCL-USER::" package when used. And these more robust if they were in the keyword package or a keyword-like package, and that would apply to the keywords in ANSI CL Loop.

**Comment by Anzus** [comment]
Hmm. There seems to be a file missing from our CLtL2 repository, I did find it on Bob’s tape dump.
